test/extended/builds/valuefrom.go

```diff
@@ -0,0 +1,210 @@
+
+		g.By("creating test secret")
+		err = oc.Run("create").Args("-f", secretFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test configmap")
+		err = oc.Run("create").Args("-f", configmapFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test successful build config")
+		err = oc.Run("create").Args("-f", successfulDockerBuildValueFrom).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("starting test build")
+		br, err := exutil.StartBuildAndWait(oc, "mydockertest")
+		o.Expect(err).NotTo(o.HaveOccurred())
+		br.AssertSuccess()
+
+		o.Expect(br.Logs()).To(o.ContainSubstring("\"FIELDREF_ENV\" \"mydockertest-1\""))
+		o.Expect(br.Logs()).To(o.ContainSubstring("\"CONFIGMAPKEYREF_ENV\" \"myvalue\""))
+		o.Expect(br.Logs()).To(o.ContainSubstring("\"SECRETKEYREF_ENV\" \"developer\""))
+		o.Expect(br.Logs()).To(o.ContainSubstring("\"FIELDREF_CLONE_ENV\" \"mydockertest-1\""))
+
+	})
+
+	g.It("should fail processing ResourceFieldRef and value/valueFrom in s2i build strategy environment variables", func() {
+
+		g.By("creating test image stream")
+		err := oc.Run("create").Args("-f", stiImageStreamFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test secret")
+		err = oc.Run("create").Args("-f", secretFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test configmap")
+		err = oc.Run("create").Args("-f", configmapFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test successful build config")
+		err = oc.Run("create").Args("-f", failedSTIBuildValueFrom).Execute()
+		o.Expect(err).To(o.HaveOccurred())
+
+		o.Expect(strings.Contains(fmt.Sprintf("%#v", err), "ResourceFieldRef is not valid")).To(o.BeTrue())
+		o.Expect(strings.Contains(fmt.Sprintf("%#v", err), "cannot use 'value' and 'valueFrom' together")).To(o.BeTrue())
+
+	})
+
+	g.It("should fail processing ResourceFieldRef and value/valueFrom in docker build strategy environment variables", func() {
+
+		g.By("creating test image stream")
+		err := oc.Run("create").Args("-f", dockerImageStreamFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test secret")
+		err = oc.Run("create").Args("-f", secretFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test configmap")
+		err = oc.Run("create").Args("-f", configmapFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating test successful build config")
+		err = oc.Run("create").Args("-f", failedDockerBuildValueFrom).Execute()
+		o.Expect(err).To(o.HaveOccurred())
+
+		o.Expect(strings.Contains(fmt.Sprintf("%#v", err), "ResourceFieldRef is not valid")).To(o.BeTrue())
+		o.Expect(strings.Contains(fmt.Sprintf("%#v", err), "cannot use 'value' and 'valueFrom' together")).To(o.BeTrue())
+
+	})
+
+	g.It("should fail processing reference $(ENV_VAR) to missing s2i build strategy environment variable", func() {
```

Comment:
(refers to this and subsequent test:) Kubernetes doesn't class an unresolvable reference as an error (it just leaves $(ENV_VAR) in the input).  Should we really diverge?